module/userExtension/helper.js

```diff
@@ -0,0 +1,490 @@
+											// If not, add them
+											// If they are present, do nothing
+											for (const role of newRoles) {
+												if (!currentRoles.includes(role)) {
+													existingUserProgramRoleMapping[currentRoleInfoIndex].roles.push(
+														role
+													)
+													kafkaEventPayloads.push(
+														createKafkaPayload(
+															userProfile,
+															programId,
+															role,
+															CONSTANTS.common.CREATE_EVENT_TYPE
+														)
+													)
+												}
+											}
+										} else {
+											// Create new program entry
+											addNewProgramEntry(
+												userProfile,
+												programId,
+												newRoles,
+												existingUserProgramRoleMapping,
+												kafkaEventPayloads
+											)
+										}
+									} else if (userRole.programOperation === CONSTANTS.common.REMOVE_OPERATION) {
+										if (currentRoleInfoIndex !== -1) {
+											const currentRoles =
+												existingUserProgramRoleMapping[currentRoleInfoIndex].roles
+
+											// Remove specified roles
+											const rolesToKeep = currentRoles.filter((role) => !newRoles.includes(role))
+
+											// Emit delete events for removed roles
+											const rolesToRemove = currentRoles.filter((role) => newRoles.includes(role))
+											for (const role of rolesToRemove) {
+												kafkaEventPayloads.push(
+													createKafkaPayload(
+														userProfile,
+														programId,
+														role,
+														CONSTANTS.common.DELETE_EVENT_TYPE
+													)
+												)
+											}
+
+											if (rolesToKeep.length === 0) {
+												// Remove entire program entry if no roles left
+												existingUserProgramRoleMapping.splice(currentRoleInfoIndex, 1)
+											} else {
+												// Update with remaining roles
+												existingUserProgramRoleMapping[currentRoleInfoIndex].roles = rolesToKeep
+											}
+										}
+										// If program doesn't exist, nothing to remove - no error needed
+									}
+								}
+							}
+
+							// Update user extension document
+							const updateQuery = { programRoleMapping: existingUserProgramRoleMapping }
+							user = await userExtensionsQueries.findAndUpdate(
+								{ _id: existingUser._id, tenantId: tenantAndOrgInfo.tenantId },
+								updateQuery,
+								{
+									new: true,
+									returnNewDocument: true,
+								}
+							)
+							userExtensionMap[user.userId.toString()] = user
+
+							userRole['_SYSTEM_ID'] = existingUser._id
+							userRole.status = 'Success'
+
+							userRole._kafkaEventPayloads = kafkaEventPayloads
+							aggregateKafkaEventPayloads.push(...kafkaEventPayloads)
+						}
+					} catch (error) {
+						userRole.status = error && error.message ? error.message : error
+					}
+
+					userRolesUploadedData.push(userRole)
+				}
+
+				for (let kafkaEventPayload of aggregateKafkaEventPayloads) {
```

Comment:
@borkarsaish65 while we are formating the payload in 2 place while pushing the kafkapayload can we push all info at once ? basically eventObj can be generated in one place

Reply:
done